### ISSUE TITLE
HLSL parser doesn't issue a detailed message when a type declaration fails to parse

### ISSUE THREAD

**Issue by dneto0** [KhronosGroup/glslang#2514]
Spawned from https://github.com/google/shaderc/issues/1154

Example input file foo.comp.hlsl:
```
foo
```
Compiling has the following output:
```
$ glslangValidator -D -V foo.comp.hlsl
foo.comp.hlsl
foo.comp.hlsl(1): error at column 3, HLSL parsing failed.
ERROR: 1 compilation errors.  No code generated.


SPIR-V is not generated for failed compile or link
```
There is normally a detailed error message. For example for this source file:
```
float3 main(){
  return a;
}
```
We get a detailed message about what went wrong:
```
$ glslangValidator -D -V foot.comp.hlsl
foot.comp.hlsl
ERROR: foot.comp.hlsl:2: 'a' : unknown variable 
ERROR: foot.comp.hlsl:2: 'return' : type does not match, or is not convertible to, the function's return type 
foot.comp.hlsl(2): error at column 11, HLSL parsing failed.
ERROR: 3 compilation errors.  No code generated.


SPIR-V is not generated for failed compile or link
```

**Comment by dneto0** [comment]
I think I found the spot where it goes wrong.
In HlslGrammar::acceptFullySpecifiedType there's a spot where the parser returns false but does not emit a message.
It has an exception to backout (or "recede") on a token, if it was "sample".  It should not issue a message then, but it should issue a message in other cases, I think.

I'll post a PR to see if it passes tests.

**Comment by dneto0** [comment]
No, the simple attempted fix fails horribly.
For what it's worth, this was the attempt:
```
$ git diff --cached
diff --git a/glslang/HLSL/hlslGrammar.cpp b/glslang/HLSL/hlslGrammar.cpp
index f30c6409..652afd32 100644
--- a/glslang/HLSL/hlslGrammar.cpp
+++ b/glslang/HLSL/hlslGrammar.cpp
@@ -612,8 +612,11 @@ bool HlslGrammar::acceptFullySpecifiedType(TType& type, TIntermNode*& nodeList,
         // If this is not a type, we may have inadvertently gone down a wrong path
         // by parsing "sample", which can be treated like either an identifier or a
         // qualifier.  Back it out, if we did.
-        if (qualifier.sample)
+        if (qualifier.sample) {
             recedeToken();
+        } else {
+            expected("type");
+        }
 
         return false;
     }
```

**Comment by dneto0** [comment]
Urgency for me is low because the user still gets an overall error message.  And the original sample case was very pathologically bad.

**Comment by jeremy-lunarg** [comment]
I think the least ambiguous error we can issue is that the expressions "foo" is an erroneous declaration.

**Comment by greg-lunarg** [comment]
Fixed by #2602. Here is the error message now:

/cygdrive/c/Users/Greg/dev/glslang/build/StandAlone/Debug/glslangValidator.exe -D -V foo.comp
foo.comp
ERROR: foo.comp:1: 'declaration' : Expected
foo.comp(1): error at column 3, HLSL parsing failed.
ERROR: 2 compilation errors.  No code generated.
